app/middleware/csrf.js

```diff
@@ -23,22 +32,21 @@ exports.csrfCheck = (req, res, next) => {
     return responseRouter.response(req, res, 'UNAUTHORISED')
   }
 
-  const chargeSession = session.retrieve(req, chargeId) || {}
+  const csrfSecret = cookies.getSessionCsrfSecret(req)
+    ?? session.retrieve(req, chargeId)['csrfSecret'] // TODO back compat for in-flight payments pre PP-12546, remove nullish coalescing op after merge
```

Comment:
I”m not sure this is going to be enough.

It solves the case of a user POSTing with a cookie that only has the CSRF secret inside the charge and not at the top level.

But consider the following case:

1. User hits a server that has been deployed with the new code and gets a new CSRF secret
2. User makes a POST that hits a server that’s still running the old code
3. Server running the old code gets the CSRF secret from inside the charge but it doesn’t match the newly-generated CSRF secret in step 1

I think we might have to make it so that the new top-level CSRF secret is stored within each charge too…

Reply:
I think in this case frontend will throw an unauthorised error. The user would already have started a payment on a server not running the new code. Refresh the card entry page and hit a server running the new code (and generating a new secret and token), then POST to an old server with a token that will not be valid with the top level secret. 

There’s not really anything we can do about that if the user POSTs to an old server. 

We’re resolving for what will be an incredibly small number of payments (if any depending on when the PR is merged). 

Perhaps we should accept the chance of a very small number of payments being impacted and that this is a tolerable outcome. 

Or merge this change out of hours.

Reply:
> The user would already have started a payment on a server not running the new code.

Yes, that could happen. The user starts a payment on a server running the old code (so they get a per-charge CSRF secret) then hits a server running the new code (so they get a new top-level secret but the per-charge one is left alone), then hits a server running the old code again (which only knows how to read the old per-charge secrets).

However, I think it could happen in both cases If the user started a payment on a server running the new code (so they get the top-level CSRF secret but not the per-charge one) then hits a server running the old code (which only knows how to read the per-charge secrets). Since that involves fewer hops between old and new, it seems more likely.

Put together, I think we might end up breaking a significant number of the in-flight payments during deployment.

How hard would it be to make the code that writes the new top-level CSRF secret also write the same secret to each charge? I think that would solve it.

Reply:
resolving this comment, we've discussed the approach and are adopting a phased roll out of changes to blunt some of these sharp edges